### DISCUSSION TITLE
Typing this in unified plugin (specifically compiler)

### DISCUSSION THREAD

**Discussion by senicko** [unifiedjs/unified#244] [Q&A] [answered]
Hello! I am writing a custom compiler for markdown AST. I can't figure out how to type `this` properly (to enforce that `CompileTree` is `Root` and `CompileResult` is my custom type. This is code which I have tried.

```ts
type TTS = { input: Array<string>; frontmatter: Record<string, string> };

declare module "unified" {
  interface CompileResultMap {
    TTS: TTS;
  }
}

function ttsCompiler(
  this: Processor<undefined, undefined, undefined, Root, TTS>
) {
  this.compiler = (ast) => ({ input: [], frontmatter: {} });
}

const ast = await unified()
  .use(remarkParse)
  .use(remarkMDX)
  .use(remarkFrontmatter)
  //  ^^^^^^^^^^^^^^^^^^^ this gives the following error
  .use(ttsCompiler)
  .process(rawMDX);
```

```
No overload matches this call.
  Overload 1 of 3, '(preset?: Preset | null | undefined): Processor<Root, undefined, undefined, undefined, undefined>', gave the following error.
    Type '(this: Processor<undefined, undefined, undefined, Root, TTS>) => void' has no properties in common with type 'Preset'.
  Overload 2 of 3, '(list: PluggableList): Processor<Root, undefined, undefined, undefined, undefined>', gave the following error.
    Argument of type '(this: Processor<undefined, undefined, undefined, Root, TTS>) => void' is not assignable to parameter of type 'PluggableList'.
  Overload 3 of 3, '(plugin: Plugin<[], undefined, undefined>, ...parameters: [] | [boolean]): Processor<Root, undefined, undefined, undefined, undefined>', gave the following error.
    Argument of type '(this: Processor<undefined, undefined, undefined, Root, TTS>) => void' is not assignable to parameter of type 'Plugin<[], undefined, undefined>'.
      The 'this' types of each signature are incompatible.
        Type 'Processor<undefined, undefined, undefined, undefined, undefined>' is not assignable to type 'Processor<undefined, undefined, undefined, Root, TTS>'.
          Type 'Root' is not assignable to type 'undefined'.ts(2769)
```

I assume that's because I can't type `this` like this in my `ttsCompile` function. But how am I supposed to do that? I looked into source of `remarkStringify` and they do something like this.

```js
export default function remarkStringify(options) {
  /** @type {Processor} */
  // @ts-expect-error: TS in JSDoc generates wrong types if `this` is typed regularly.
  const self = this
```

When I assign `this` to `self` in my code like that it "works" fine too (I don't get any errors). Is this the proper solution? I really appreciate any tips!

Also, can I somehow make the output to be of the `TTS` type? The outputted vfile's result is still `unknown` and I can't see any generic type in which I could pass my custom `TTS` type. Is this even possible or I just need to rely on casts?

**Comment by wooorm** [comment]
Hey!

Use `Plugin<Parameters, string, TheTypeYouWant>` to type your plugin that adds a particular compiler.
That should do the trick.

This is a bit weird because:

a) the `this` you want in the plugin could actually be *any* current unified processor; it could have some compiler already; your plugin is overwriting it. So we can’t “get” which output you intend to compile to from a particular `this`

b) TS for JS is buggy with the `this` parameter: IIRC it was that it “works” in the file but results in different `.d.ts` files.

> Also, can I somehow make the output to be of the TTS type? The outputted vfile's result is still unknown and I can't see any generic type in which I could pass my custom TTS type. Is this even possible or I just need to rely on casts?


IIRC what you want should indeed happen.

**Reply by senicko** [reply]
Hmm, I still have some issues with this approach. This is my current code

```ts
export type CompiledArticle = {
  sections: Array<string>;
  frontmatter: ArticleFrontmatter;
};
```

```ts
const ttsCompiler: Plugin<[], Root, CompiledArticle> = function () {
  this.compiler = (ast, file) => {
    const frontmatter = file.data.frontmatter as ArticleFrontmatter;
    const sections = [frontmatter.title + "\n\n"];

    for (const child of ast.children) {
      //                    ^^^^^^^^ <------------------------ this fails as ast has type unist.Node and not mdast.Root
      const text = toString(child) + "\n\n";

      if (sections[sections.length - 1].length + text.length <= 4096) {
        sections[sections.length - 1] += text;
      } else {
        sections.push(text);
      }
    }

    return { sections, frontmatter };
  };
};
```

```
Property 'children' does not exist on type 'Node'.
```

Is there any way to say "I want a compiler that receives mdast.Root and returns CompiledArticle"?

**Reply by ChristianMurphy** [reply]
Heya @senicko! 👋 
A few pointers

---

```ts
const frontmatter = file.data.frontmatter as ArticleFrontmatter;
```

I'd highly discourage using a force cast on a type, authors may not perfectly capture the json/yml in their front matter.
Consider using a type validator like https://zod.dev/ to ensure the shape of the data is what you expect, and give clear error messaging if it is not.

---

```ts
for (const child of ast.children) {
      //                    ^^^^^^^^ <------------------------ this fails as ast has type unist.Node and not mdast.Root
```

This is accurate, a compiler is likely, but not guaranteed to receive a root node.
You'll want assert here using the ideas from https://unifiedjs.com/learn/recipe/narrow-node-typescript/
Most likely something like

```ts
import type {Root} from 'mdast'
import {is} from 'unist-util-is'

// ...

if (is<Root>(ast, 'root')) {
  // ast is known to be root in this context
  for (const child of ast.children) {
    // do your logic
  }
} else {
  // do some error handling
}
```

**Reply by senicko** [reply]
1) Fully agreed, I will use zod

2) Actually, it is not possible to do this. I have opened separate discussion thread in `syntax-tree` https://github.com/orgs/syntax-tree/discussions/128#discussioncomment-9618585 where you advised me to write plugins for my logic. When you pass mdast.Root as a generic argument to is you can no longer use "root" as a Test. I am probably not understanding something here 😟 

I hope the snippet has everything that's relevant to this problem ;d

```ts
import type { Root } from "mdast";
import { is } from "unist-util-is";

declare module "unified" {
  interface CompileResultMap {
    // Register a new result (value is used, key should match it).
    CompiledArticle: CompiledArticle;
  }
}

const ttsCompiler: Plugin<[], Root, CompiledArticle> = function () {
  this.compiler = (ast, file) => {
    if (is<Root>(ast, "root")) {
      //             ^^^^^^^^ (1)
      for (const child of ast.children) {
        //                ^^^^^^^^^^^^ (2)
      }

      return { sections: [], frontmatter: { title: "", slug: "" } };
    } else {
      throw new Error("ttsCompiler can be run only on mdast.Root");
    }
  };
};

/** compileArticle extract frontmatter and ttsInput from an article. */
export async function compileArticle(rawMDX: string): Promise<CompiledArticle> {
  const vfile = await unified()
    .use(remarkParse)
    .use(remarkMDX)
    .use(remarkFrontmatter)
    .use(ttsTransformer)
    .use(remarkUnlink)
    .use(ttsCompiler)
    .process(rawMDX);

  return vfile.result;
}
```

(1):
```
Argument of type 'string' is not assignable to parameter of type 'Root'
```

(2):
```
Property 'children' does not exist on type 'Node'.
```

I really appreciate your help. I would like to get this working so much 😆

**Comment by wooorm** [comment]
Why are you even trying to create a custom compiler that returns a non-string result?
In your compiler, you access and later return `file.data.frontmatter`.
Anyone outside of your customization can already access `file.data.frontmatter`.

> Is there any way to say "I want a compiler that receives mdast.Root and returns CompiledArticle"?

Yes, by typing your `compiler` function too.
Type both the plugin **and** `compiler`

**Reply by senicko** [reply]
I am trying to create a custom compiler that returns a non-string result because I want to read my MDX articles with OpenAI's tts. The API accepts input of at most 4096 characters. I am trying to write a tool which takes my markdown files and turns them into an array of "sections" of at most 4096 characters. That's why I want to compile my markdown file into an array of strings, which I can later pass to the tts model.

My article's title is stored in frontmatter, that's why I access it in my compiler, to append the title to the first section. I need the frontmatter later, in other parts of the tool, for example to name .mp3 files correctly, so I return it together with compilation result (I didn't want to return vfile to `compileArticle` caller). I didn't finish writing this code, if I will figure out how to make the types work maybe I will do something different with the frontmatter, I didn't ask about it.

> by typing the compiler function too

That's what I want to do! And I don't know how. I don't want to be annoying but I didn't create or worked on unified, so I have no idea how to do this and I couldn't find any docs on that or figure out how the types are passed around internally. Maybe that's a skill issue, that's why I am asking.

- `unified()` itself does not accept any generics. 

- When I type `this` to be `Processor<undefined, undefined, undefined, Root, CompiledArticle>` I can no longer assign this function to `const ttsCompiler: Plugin<[], Root, CompiledArticle>`.

```ts
const ttsCompiler: Plugin<[], Root, CompiledArticle> = function (this: Processor<undefined, undefined, undefined, Root, CompiledArticle>) {
```

```
Type '(this: Processor<undefined, undefined, undefined, Root, CompiledArticle>) => void' is not assignable to type 'Plugin<[], Root, CompiledArticle>'.
  The 'this' types of each signature are incompatible.
    Type 'Processor<undefined, undefined, undefined, undefined, undefined>' is not assignable to type 'Processor<undefined, undefined, undefined, Root, CompiledArticle>'.
      Type 'Root' is not assignable to type 'undefined'.
```

- I can't just change the type of `ast`

```
  this.compiler = (ast: Root, file) => {
    //             ^^^
```

```
Type '(ast: Root, file: VFile) => { sections: string[]; frontmatter: { title: string; slug: string; }; }' is not assignable to type 'Compiler<Node, CompileResults>'.
  Types of parameters 'ast' and 'tree' are incompatible.
    Property 'children' is missing in type 'Node' but required in type 'Root'.
```

- I would love to do what @ChristianMurphy advised, but it doesn't work (in my case, I would love to make it work, this approach is even shown in recipes articles!). You can't pass generic `Root` to `is` and use `string` test as the generic type is used as condition. That's what internal typings do.

`unist-util-is/lib/index.d.ts` line 63
```ts
export const is: (<Condition extends string>(
  node: unknown,
  test: Condition,
  index?: number | null | undefined,
  parent?: Parent | null | undefined,
  context?: unknown
) => // ...
```

How am I supposed to do it then? It's just a casual question. If it's not possible it's also fine, It's just about types, I can cast some things around.

**Reply by senicko** [reply]
Sure, If it doesn't work I will just stringify the AST, and split it into "sections" outside of unified, but I just decided to try to do all of this inside of a compiler.

**Reply by wooorm** [reply]
> That's what I want to do! And I don't know how. 

How do you mean, I find it hard to imagine you don’t know how to type a function?

```js
this.compiler = compiler

function compiler(ast: Root, file: VFile): CompiledArticle {
 // ...
}
```

**Reply by senicko** [reply]
I thought you've meant something different 😄 This, as I have mentioned, doesn't work. Even if I declare a separate compiler function.

```ts
const ttsCompiler: Plugin<[], Root, CompiledArticle> = function () {
  this.compiler = compiler;

  function compiler(ast: Root, file: VFile): CompiledArticle {
    return { sections: [], frontmatter: { title: "", slug: "" } };
  }
};
```

```
Type '(ast: Root, file: VFile) => CompiledArticle' is not assignable to type 'Compiler<Node, CompileResults>'.
  Types of parameters 'ast' and 'tree' are incompatible.
    Property 'children' is missing in type 'Node' but required in type 'Root'.
```

**Reply by wooorm** [reply]
> I didn't ask about it.

Often, the answers humans are looking for, are not the answers to the questions they ask.

You‘re doing something there that is likely a bad idea, so I point it out, to prevent you from running into more problems.

> When I type this to be Processor<undefined, undefined, undefined, Root, CompiledArticle> I can no longer assign this function to const ttsCompiler: Plugin<[], Root, CompiledArticle>.

Yes, this is what I tried to point out before: *do not type `this` like that*. You type your `compiler` function as a very specific thing. And you type your plugin as a very specific thing. You leave `this` as vague as possible. Because people could use your specific plugin with all kinds of different processors. Your plugin is *specific*. `this` is vague.

**Reply by wooorm** [reply]
Something along your last lines should work. If it doesn’t work, make a reproduction. You use a lot of variables and I don’t know what they refer to.

**Reply by senicko** [reply]
This is the reproduction which still has the error.

(edit): I've removed the `compileArticle` function from previous snippets as it does not change anything in terms of the produced type error.

```ts
import type { Root } from "mdast";
import remarkMDX from "remark-mdx";
import remarkParse from "remark-parse";
import { unified, type Plugin } from "unified";
import type { VFile } from "vfile";

export type CompiledArticle = {
  sections: Array<string>;
};

declare module "unified" {
  interface CompileResultMap {
    // Register a new result (value is used, key should match it).
    CompiledArticle: CompiledArticle;
  }
}

const ttsCompiler: Plugin<[], Root, CompiledArticle> = function () {
  this.compiler = compiler;

  function compiler(ast: Root, file: VFile): CompiledArticle {
    return { sections: [] };
  }
};

const rawMDX = `
import Caption from "../../components/Caption.astro";
import CodeSnippet from "../../components/CodeSnippet.astro";

## Introduction
`;

const vfile = await unified()
  .use(remarkParse)
  .use(remarkMDX)
  .use(ttsCompiler)
  .process(rawMDX);

console.log(vfile.result);
```

```
Type '(ast: Root, file: VFile) => CompiledArticle' is not assignable to type 'Compiler<Node, CompileResults>'.
  Types of parameters 'ast' and 'tree' are incompatible.
    Property 'children' is missing in type 'Node' but required in type 'Root'.
```

**Reply by wooorm** [reply]
repro needs package.json + tsconfig + filenames + commands for me to be able to repro

**Reply by senicko** [reply]
@wooorm here you go
https://github.com/senicko/unified-compiler-plugin-reproduction

**Reply by wooorm** [reply, marked as answer]
Using `index.ts` and latest packages, VS code does show that type error.
I recommend adding at `@ts-expect-error`:

```diff
const ttsCompiler: Plugin<[], Root, CompiledArticle> = function () {
+ // @ts-expect-error
  this.compiler = compiler;
```

Or cast:

```diff
- function compiler(ast: Root, file: VFile): CompiledArticle {
+ function compiler(ast: Node, file: VFile): CompiledArticle {
+   const tree = ast as Root
    return { sections: [] };
  }
```

The `unified` types seem to tie a) what the current compiler could be, with b) the new compiler intended to be set.

**Reply by senicko** [reply]
Okay, thanks a lot. @ChristianMurphy's approach seems to be the most "valid" one, but I feel like `is` type is broken (? maybe it works like it's supposed to, but certainly doesn't allow to narrow ts type in the way it's show in recipe, or I am missing something). Whatever, I really appreciate your help @wooorm @ChristianMurphy 

Have a nice day :)
